presto-spark-launcher/src/main/java/com/facebook/presto/spark/launcher/PrestoSparkRunner.java

```diff
@@ -122,48 +134,64 @@ private static IPrestoSparkServiceFactory createServiceFactory(File directory)
         return serviceLoader.iterator().next();
     }
 
-    private static class CachingServiceFactory
-            implements Serializable
+    private static IPrestoSparkService createService(
+            SparkProcessType sparkProcessType,
+            PackageSupplier packageSupplier,
+            Map<String, String> configProperties,
+            Map<String, Map<String, String>> catalogProperties)
     {
-        private static IPrestoSparkService service;
+        String packagePath = getPackagePath(packageSupplier);
+        File pluginsDirectory = checkDirectory(new File(packagePath, "plugin"));
+        PrestoSparkConfiguration configuration = new PrestoSparkConfiguration(configProperties, pluginsDirectory.getAbsolutePath(), catalogProperties);
+        IPrestoSparkServiceFactory serviceFactory = createServiceFactory(checkDirectory(new File(packagePath, "lib")));
+        return serviceFactory.createService(sparkProcessType, configuration);
+    }
 
-        private static String currentPackagePath;
-        private static Map<String, String> currentConfigProperties;
-        private static Map<String, Map<String, String>> currentCatalogProperties;
-        private static SparkProcessType currentProcessEnvironment;
+    private static String getPackagePath(PackageSupplier packageSupplier)
+    {
+        return checkDirectory(packageSupplier.getPrestoSparkPackageDirectory()).getAbsolutePath();
+    }
 
+    private static class DistributionBasedPrestoSparkTaskExecutorFactoryProvider
```

Comment:
This class name is very wordy.. 😃

Reply:
Yeah, unfortunately i didn't find a better name. We can think of a better name later.